### PR TITLE
UCP/DOC: Request init/cleanup callbacks doc clarification

### PR DESCRIPTION
**Pull request by yosefe**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5953/ for details (Mellanox internal link).

File: src/ucp/api/ucp.h [hunk 1]
```diff
@@ -582,16 +582,24 @@ typedef struct ucp_params {
...
      */
     ucp_request_init_callback_t        request_init;
 
     /**
-     * Pointer to a routine that is responsible for cleanup the memory
-     * associated with the request.  @e NULL can be used if no such function
-     * required.
-     * This field defaults to @e NULL if not specified.
+     * Pointer to a routine that is responsible for final cleanup of the memory
+     * associated with the request. This will routine will not be called every
```
Comment (brminich) [inline src/ucp/api/ucp.h]: `This will routine` -> `This routine`

**Comment by yosefe** [comment]
@shamisp @bbenton pls take a look

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5955/ for details (Mellanox internal link).

File: src/ucp/api/ucp.h [hunk 2]
```diff
@@ -582,16 +582,24 @@ typedef struct ucp_params {
...
      */
     ucp_request_init_callback_t        request_init;
 
     /**
-     * Pointer to a routine that is responsible for cleanup the memory
-     * associated with the request.  @e NULL can be used if no such function
-     * required.
-     * This field defaults to @e NULL if not specified.
+     * Pointer to a routine that is responsible for final cleanup of the memory
+     * associated with the request. This routine will not be called every time a
```
Comment (shamisp) [inline src/ucp/api/ucp.h]: will not be called -> may not be called

File: src/ucp/api/ucp.h [hunk 3]
```diff
@@ -582,16 +582,24 @@ typedef struct ucp_params {
...
+     * request is released; instead, it will be called only once during
```
Comment (shamisp) [inline src/ucp/api/ucp.h]: For some implementation the cleanup call maybe delayed and only invoked at ucp_worker_cleanup

**Comment by yosefe** [comment]
@shamisp fixed

**Comment by shamisp** [comment]
👍

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5979/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@bbenton pls take a look

**Comment by yosefe** [comment]
@bbenton @gmegan can you pls review?

File: src/ucp/api/ucp.h [hunk 4]
```diff
@@ -582,16 +582,24 @@ typedef struct ucp_params {
...
     /**
      * Pointer to a routine that is used for the request initialization.
-     * @e NULL can be used if no such function required.
-     * This field defaults to @e NULL if not specified.
+     * This function will be called only on the very first time a request memory
+     * is initialized, and may not be called again if a request is reused.
+     * If a request should be reset before the next reuse, it can be done before
+     * calling @ref ucp_request_free.
+     *
+     * @e NULL can be used if no such function required, which is also the default
```
Comment (bbenton) [inline src/ucp/api/ucp.h]: s/such function required/such function is required/

File: src/ucp/api/ucp.h [hunk 5]
```diff
@@ -582,16 +582,24 @@ typedef struct ucp_params {
...
-     * Pointer to a routine that is responsible for cleanup the memory
-     * associated with the request.  @e NULL can be used if no such function
-     * required.
-     * This field defaults to @e NULL if not specified.
+     * Pointer to a routine that is responsible for final cleanup of the memory
+     * associated with the request. This routine may not be called every time a
+     * request is released. For some implementations, the cleanup call may be
+     * delayed and only invoked at @ref ucp_worker_cleanup.
+     *
+     * @e NULL can be used if no such function required, which is also the default
```
Comment (bbenton) [inline src/ucp/api/ucp.h]: s/such function required/such function is required/

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6073/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@bbenton fixed

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6101/ for details (Mellanox internal link).

**Comment by bbenton** [comment]
:+1: (better late than never)

**Pull request merged**